server/src/main/java/io/crate/expression/scalar/KnnMatch.java

```diff
@@ -0,0 +1,174 @@
+    public KnnMatch(Signature signature, BoundSignature boundSignature) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+        this.conf = new IndexWriterConfig(new StandardAnalyzer());
+        this.type = (DataType<float[]>) boundSignature.argTypes().get(0);
+    }
+
+    @Override
+    public Signature signature() {
+        return signature;
+    }
+
+    @Override
+    public BoundSignature boundSignature() {
+        return boundSignature;
+    }
+
+    @Override
+    @SafeVarargs
+    public final Boolean evaluate(TransactionContext txnCtx,
+                                  NodeContext nodeContext,
+                                  Input<Object>... args) {
+        Object field = args[0].value();
+        if (field == null) {
+            return null;
+        }
+        Object target = args[1].value();
+        if (target == null) {
+            return null;
+        }
+        Object k = args[2].value();
+        if (k == null) {
+            return null;
+        }
+        assert field instanceof float[] : "First parameter value must be a float array";
+        assert target instanceof float[] : "Second parameter value must be a float array";
+        assert k instanceof Integer : "Third parameter value must be an integer";
+
+        try {
+            return evaluate((float[]) field, (float[]) target, (int) k);
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private boolean evaluate(float[] field, float[] target, int k) throws IOException {
+        if (field.length != target.length) {
+            throw new IllegalArgumentException("Field dimensions must match target dimensions for knn_query");
+        }
+        if (field.length != type.characterMaximumLength()) {
+            throw new IllegalArgumentException("Field dimensions must match type dimensions for knn_query");
+        }
+        try (ByteBuffersDirectory directory = new ByteBuffersDirectory()) {
+            try (IndexWriter indexWriter = new IndexWriter(directory, conf)) {
+                float[] fieldValues = (float[]) field;
+                FieldType fieldType = new FieldType(FloatVectorFieldMapper.Defaults.FIELD_TYPE);
+                fieldType.setVectorAttributes(
+                    type.characterMaximumLength(),
+                    VectorEncoding.FLOAT32,
+                    VectorSimilarityFunction.EUCLIDEAN
+                );
+                var vectorField = new KnnFloatVectorField("dummy", fieldValues, fieldType);
+                indexWriter.addDocument(List.of(vectorField));
+                indexWriter.commit();
+
+                IndexSearcher indexSearcher = new IndexSearcher(DirectoryReader.open(directory));
+                TopDocs result = indexSearcher.search(new KnnFloatVectorQuery("dummy", target, k), 1);
+                return result.scoreDocs.length > 0 && result.scoreDocs[0].score >= DEFAULT_MIN_SCORE;
+            }
+        }
```

Comment:
Is it alternative to Lucene query. ie body of the `GenericFunctionQuery`?

If so,  have 2 comments:

1. As mentioned in another comment, Lucene query probably has Euclidean as default. 
Can we add a comment next to  `VectorSimilarityFunction.EUCLIDEAN` that it's aligned with Lucene query default? In case we ever support other types, we can also adjust here. Minor thing for a code reader.

2. We are creating quite some amount of objects per-row. I looked into the classes, looks like no chance to reuse neither of them? Since we doing Euclidean (now and probably always?) I wonder if it would be faster just to do it on or own i.e just 
sqrt ((field1-target1)^2 + ... + (fieldN-targetN)^2). 

I mean it makes sense to rely on "library" implementation when it falls into Lucene query and probably involves some internal infrusture/optimizations how it accesses indices but here we already have 2 vector in our hands and  creating "library" object which eventually does sqrt... thing. 

Not in this PR of course but maybe add a comment or TODO that we can benchmark here?

Reply:
Pushed a fixup

Reply:
>My plan was to do another follow up and see if we can make this configurable via a type parameter.

I think if we support more than only Euclidean than usage of "library" objects cannot be replaced/adds too much complexity? This comment of doing `GenericFunctionQuery` on our own was only about Euclidean.